src/canister_tests/src/tests.rs

```diff
@@ -66,6 +66,462 @@ fn registration_with_mismatched_sender_fails() {
+            principal_1(),
+            user_number,
+            frontend_hostname.to_string(),
+            pub_session_key.clone(),
+            expiration,
+        )? {
+            GetDelegationResponse::SignedDelegation(delegation) => delegation,
+            GetDelegationResponse::NoSuchDelegation => panic!("failed to get delegation"),
+        };
+
+        framework::verify_delegation(&env, canister_sig_key, &signed_delegation);
+        assert_eq!(signed_delegation.delegation.pubkey, pub_session_key);
+        assert_eq!(signed_delegation.delegation.expiration, expiration);
+        Ok(())
+    }
+
+    /// Verifies that delegations can be requested in parallel.
+    #[test]
+    fn should_get_multiple_valid_delegations() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+        let user_number = flows::register_anchor(&env, canister_id);
+        let frontend_hostname_1 = "https://dapp1.com";
+        let frontend_hostname_2 = "https://dapp2.com";
+        let pub_session_key_1 = ByteBuf::from("session public key 1");
+        let pub_session_key_2 = ByteBuf::from("session public key 2");
+        let delegation_params = vec![
+            (
+                &pub_session_key_1,
+                frontend_hostname_1,
+                SystemTime::UNIX_EPOCH,
+            ),
+            (
+                &pub_session_key_1,
+                frontend_hostname_2,
+                SystemTime::UNIX_EPOCH,
+            ),
+            (
+                &pub_session_key_2,
+                frontend_hostname_1,
+                SystemTime::UNIX_EPOCH,
+            ),
+            (
+                &pub_session_key_1,
+                frontend_hostname_1,
+                SystemTime::UNIX_EPOCH.add(Duration::from_secs(30)),
+            ),
+        ];
+
+        let mut prepare_delegation_results = vec![];
+        for (session_key, frontend_hostname, time) in delegation_params.clone() {
+            env.set_time(time);
+            let (canister_sig_key, expiration) = api::prepare_delegation(
+                &env,
+                canister_id,
+                principal_1(),
+                user_number,
+                frontend_hostname.to_string(),
+                session_key.clone(),
+                None,
+            )?;
+
+            assert_eq!(
+                expiration,
+                env.time()
+                    .add(Duration::from_secs(30 * 60)) // default expiration: 30 minutes
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_nanos() as u64
+            );
+            prepare_delegation_results.push((canister_sig_key, expiration))
+        }
+
+        for (idx, (session_key, frontend_hostname, _)) in delegation_params.iter().enumerate() {
+            let (canister_sig_key, expiration) = prepare_delegation_results.get(idx).unwrap();
```

Comment:
question: how do we make sure the delegation is ready here? when does the signing/certification happen?

Reply:
The `state_machine` executes ingress synchronously, i.e. the `prepare_delegation` update call will only return once it is complete. The whole `call` and then `read_state` polling has been cut away in this test infrastructure.

Reply:
gotcha, a short comment then would help